### DISCUSSION TITLE
"Downsampling" flows

### DISCUSSION THREAD

**Discussion by zxyz** [akvorado/akvorado#1527] [Q&A] [answered]
Hey there,
from my experience flow data is used for three things:

- forensic analysis (very low sampling rate)
- ddos detection (low sampling rate)
- observability (higher sampling rate)

From an observability standpoint the sample rate does not need to be very low [1], e. g. for peering analysis trends are mostly sufficient to know whom to ask for a PNI or for analysing which kinda data is filling up the backbone links.

For DDoS detection to not miss smaller DDoS attacks a lower sampling rate than for observability is needed. And for forensic analysis after an e. g. extortion/ransomware attack an even lower sampling rates come in handy.

As long as the overall Flows per second rate is not very high using a low sampling rate is probably working for all above use cases. But with several million FPS ingestion gets expensive (hardware and power consumption wise). Most flow sampling with network devices is done in hardware and only one sampling rate is possible (e. g. not possible to sample with 1:1000 to DDoS detector and sample with 1:8000 to akvorado, at least I never came across a vendor where this was possible).

I was thinking of using a replicator (nfdump, pmacct or alike) and forward only every X'th packet from each exporter to akvorado. Unfortunately I was not able to find existing software which is able to do this (anyone knows any?). So I was thinking if it might be possible to do this in the akvorado inlet (?)?

People suggested to use a aggregating merge tree in clickhouse using the right sampling key, but at that point the data was already processed and written to disk and akvorado `resolutions` would probably do the trick.

Any thought on how to solve this? Thanks a lot and have a great weekend.

[1] https://kb.kentik.com/v0/Ab02.htm#Ab02-Optimal_Sampling_Rates

**Comment by vincentbernat** [comment, marked as answer]
There is a `rate-limit` setting in the inlet. https://demo.akvorado.net/docs/configuration#flow

**Reply by zxyz** [reply]
awesome! thanks a lot for pointing this out (missed it in the documentation).

**Comment by mateuszbm** [comment]
Is rate-limit still available? Can't find it in documentation.

**Comment by zxyz** [comment]
it was removed with 2.0:

> inlet: flow rate limiting feature has been removed

https://demo.akvorado.net/docs/changelog#200---2025-09-22

**Reply by mateuszbm** [reply]
Do you found  another way to workaround removed feature?

**Reply by zxyz** [reply]
did not look for one.

**Reply by vincentbernat** [reply]
It may be added back, but it will be a bit differently: the sampling would be done in the outlet, not the inlet. So, the flows will be temporarily stored in Kafka and costing a bit of CPU.

**Reply by mateuszbm** [reply]
Would love to give it a try if it is possible.
